Initial pull request to establish a baseline

Zendog opened the pull request:
This version provide the following fixes and upgrades:

Updated soco library to - 0.30.9 for better control function support and overall stability.
Rewrote and fixed subscription services, includes fallback subscription upon failure. - Should fix the reported subscription issues.
Rewrote and corrected volume control. - will fix the reported volume control issues.
Rewrote and corrected Bass and Treble control. - added functionality
Implemented SiriusXM login and channel processing. - Fixes reported XM issues.
Rewrote and corrected Pandora loging / crypto processing from python 2 to python 3 based calls. Fixes non reported issues.
Implemented Pandora login and channel processing.
Removed excess modules from requirements.txt

Zendog committed:
Update requirements.txt

Zendog committed:
Added module display to startup

Added module display to startup to identify this plugin as being executed for support and version control.

==================================================================================================================================================================
Plugin Name:                   Sonos_NG
Plugin Version:                2024.0.9
Plugin ID:                     com.ssi.indigoplugin.trasente.Sonos
Indigo Version:                2024.2.0
Indigo License:                ActiveSubscription
Indigo API Version:            3.6
Architecture:                  arm64
Python Version:                3.11.6
Mac OS Version:                15.3.2
Plugin Process ID:             70564
SoCo Version:                  0.30.9
SoCo Path:                     /Library/Application Support/Perceptive Automation/Indigo 2024.2/Plugins/Sonos.indigoPlugin/Contents/Server Plugin/soco/__init__.py
==================================================================================================================================================================

Zendog committed:
Updated requirements.txt

Updates for additional modules to be installed.

Zendog committed:
Added additional requirements

# Core dependencies
soco==0.30.9
lxml==5.1.0
requests>=2.31.0
netifaces==3.1.1
ifaddr==0.2.0
xmltodict==0.13.0

# Audio & TTS support
mutagen==1.47.0
gTTS==2.5.0
pyvona==1.2.1

# AWS support for Polly (and general use)
boto3==1.34.9
botocore==1.34.9

# Twisted networking (used by SoCo events, SSDP)
Twisted==23.10.0

# Optional: HTTP/SSL support
urllib3==2.0.7

Zendog committed:
Corrected Install Compiled lxml Moduls

Corrected lxml compiled C library that appears to be needed in compiled form during installation.

Zendog committed:
Changed gitignore to allow compiled C modules

Changed .gitignore file to allow compiled C modules to traverse the repository.

Zendog committed:
📦 Force-include lxml and pandora.py for full plugin packaging

Zendog committed:
Customized .gitignore

Customized .gitignore to be plugin safe and include all needed modules for processing.

Zendog committed:
Added Info black to startup

Added info block to ensure soco version and plugin version that is running.

Zendog committed:
Removed binaries

Test loading with removed binaries.

Zendog committed:
Include Pandora

Added pandora install

Zendog committed:
remove packages

remove test

Zendog committed:
Remove external pandora test

test remove

Zendog committed:
Latest Version

Updated

FlyingDiver commented:
All those libraries should be removed from the plugin and listed in requirements.txt for inclusion at install time.

Some of those libraries conflict with ones that are already installed by Indigo.

Zendog committed:
Test remove libraries

Test remove libraries

Zendog committed:
Fix soco version number

Fix soco version number

Zendog committed:
Fixed urllib version

Fixed urllib version

Zendog committed:
Fixed xmltodict version

Fixed xmltodict version

Zendog committed:
Removed Libraries

Removed Libraries

Zendog committed:
Constantly version

Constantly version

Zendog committed:
Removed constantly

Removed constantly

FlyingDiver commented:
Indigo includes requests already, try it without that one.

Zendog committed:
Remove lame

Remove lame

Zendog committed:
Removed requests

Removed requests

Zendog commented:
Removed excess libraries ....

Ghawken commented:
Hi,

Well done on updating and Good luck!   Replying quickly before head out...

I tried installing (your latest commit here from your githuib) but still missing a lot of libraries, and getting error to install.  (netifaces initially, but when I manually added, soco not found - not sure why on a brief look)

You need to make your system only use the requirements.txt install libraries.

I would suggest you pip uninstall all the ones (from non Packages) you wish to use in plugin.

You can list them (adding whatever ones I'm missing here)

```
/Library/Frameworks/Python.framework/Versions/3.11/bin/python3 \
  -m pip list | grep -E 'soco|lxml|netifaces|ifaddr|xmltodict|mutagen|gTTS|pyvona|boto3|botocore|Twisted'
```

But really need to use something like below -  /Library/Frameworks/Python.framework/Versions/3.11/bin/pip3  - should be pathway to Indigo python3 install - incase have multiple.

Add add whatever other libraries are needed to the uninstall, and then will have a system that is using the requirements.txt alone, and the /Packages directory...

```
sudo /Library/Frameworks/Python.framework/Versions/3.11/bin/pip3 uninstall -y \
  soco netifaces ifaddr xmltodict \
  mutagen gTTS pyvona boto3 botocore Twisted
```

indigo-jay commented:
I agree with Glenn. One issue though - Indigo installs some of those, so you wouldn't want to uninstall them. The [list of what Indigo installs](https://wiki.indigodomo.com/doku.php?id=indigo_2024.2_documentation:python_packages#python_3116_packages_installed_by_indigo) is on the page linked on the forum topic. And, of course, you don't need to put those in your requirements.txt since we install them automatically.

Zendog committed:
Updated requirements.txt

Added fake xm and fake-useragent for xm id processing.

Zendog committed:
Removed SX client

Removed sxm client since it is embedded.

Zendog committed:
Update requirements.txt

Added additional xm helper packages.

Zendog commented:
I am perplexed as to why the list (soco|lxml|netifaces|ifaddr|xmltodict|mutagen|gTTS|pyvona|boto3|botocore|Twisted) would be barking at this point, as these appear to be installing now with the latest version of requirements.txt / or already installed as base Indigo? I can see these in the installed plugin after processing whereas they were not in the plugin pre-installation.

Others seem to have installed the latest version and these did not bark for them? I am still chasing a few dogs and cats from the other embedded routines so the uninstall suggestion is a sound one. 

To be honest, this is starting to turn into more of a goose chase for me than I was anticipating. This started out to be a simple correction for use at home (and it is working great). Making it is distributable is turning out to be more painful than the actual  fixes and enhancements ....

Zendog commented:
I was able to confirm successful installation on a fresh Indigo box at a second location that never had this plugin installed before. The installation was successful without error. So I am thinking this is working now. There is only one Sonos Playbar at that location and I am controlling it remotely, so while I cannot physically verify the sound, it appears to be controlling things fine from all indications.

indigo-jay commented:
Sounds good. I'm waiting for Github to finish "processing" (whatever that is) this PR before we merge it. Once it's done, we'll merge and create a new release. Thanks a lot for the patience working through getting this plugin updated.

Zendog committed:
SiriusXM channel number from ZP_STATION,

Update PlayBar only channel processing for device group controls inherent on Playbar devices. Updated requirements modules to support.

Zendog commented:
May be best to hold off on the merge. After testing further on a few more boxes this afternoon, I am seeing a pattern. It appears that if the machine being installed upon is fairly modern and has the capability to compile C, then the install goes well. If however, the box in question cannot compile for any reason, lack of Xcode, missing C headers, etc. then the installs can fail when attempting to compile the binaries.

So long story short, I may need to reevaluate the inventory of functions and see if there are less install intensive options that would be more compatible with the supported distribution methods. I suspect if propagated in its current form, this would lead to very inconsistent user experiences.

I am out of town business this week so it may be a while before I can address this further. I will provide an update when I have better path to offer. More to follow .....

Zendog committed:
Rewrote Interfaces

Rewrote Interfaces detection to avoid compiled module requirements.

indigo-jay commented:
This can definitely be an issue. The libraries we install use a flag to force install of binary wheels only - so it will install the most recent version from pypi that has a binary distribution (if the library contains compiled code). I don't believe you can do that in a requirements file, but what you probably can do is identify the library on pypi and see if you can identify a version that has a binary wheel - then you specify that version in the requirements.txt file and it **should** install without needing to compile.

Now, there is a lot to say about simplifying the code by removing stuff that can be replaced easily. I believe this plugin started as a python 2 plugin so it's very likely that python 3 will offer some features that can be used in place of those libraries. There are a large number of libraries specified - I suspect many can be replaced/removed.

Also on a personal note, I have to say I'm really not a fan of the emoji's in the log messages. Back when this plugin was built, we didn't color log lines in the Event Log window so I guess maybe it made a little more sense. We color code log lines now which makes it much easier to see the difference between levels, so they are redundant. And, if you use an RSS reader to follow the forums, the emoji's end up getting blown up REALLY BIG. Which makes reading posts exceedingly difficult when you are constantly scrolling past huge emojis. Not saying that you should remove them, but as a warning to others that might be tempted that it's really not a great idea...

Zendog committed:
Update Sonos.py

Lets use config panel to define subnet.

Zendog commented:
After a quick inventory check, it appears the compiled stuff was only used in a few initial discovery functions. Was easily replaced with non-compiled functionality. I did add a config option to define the most likely Sonos subnet that allows overall reduced network discovery. The latest staged GitHub versions should correct the install requirements and eliminate any compile / wheel requirement kind of things.

Zendog committed:
Eliminated Duplicated Block

Eliminated Duplicated Block in Plugin.py

Ghawken commented:
Okay.
New version installs okay on my system with the requirements.txt file as per last commit
Yah!

But you have changed the 
Plugin Identifier?
to:
`	<string>com.ssi.indigoplugin.trasente.Sonos</string>`

Which means all devices/control pages/triggers etc. will not longer exist/work... 

Not sure the politics of this, but regardless will be a bit of user-pain to change update all aspects and if doing so probably need a new Plugin Name?

As it seems the plugin Name is still Sonos - so I have one Sonos in Plugins - but when create devices, have the ability to choose between two Sonos plugins.  Plugin folders also show Sonos.indigoplugin in both Disabled and enabled folders...

@DaveL17  Could keep the identifiers the same here?  Is that plan.

Thanks again

DaveL17 commented:
It's usually best to leave the identifier the same (original) to avoid users having to rebuild their setups.

Ghawken commented:
I would agree.  I suspect the name the same name (Sonos) but change in identifier is causing some of the forum issues…
(the older Sonos plugin is still floating around, even running somehow - although doesn’t appear in the plugin list, presumably that is name sorted/selected)

DaveL17 commented:
I believe the name can change without issue, but the identifier is the main way that Indigo differentiates plugins and their resources. I recall doing a hack years ago where I needed to have two instances of the same plugin running simultaneously and was able to achieve that by using two instance with different identifiers.

Another thing that can cause issues is using symlinks in development (a lot of us do). It's important to manage them carefully as I think the OS will honor the symlink even if the target is in the recycle bin.

Zendog committed:
Update Info.plist

Changed the name back to see if this resolves apparent installation issues where Indigo is not releasing old plugin version modules during installation of the new plugin.

Zendog commented:
Hmmm....  Interesting .... I changed it to be able to separate and ensure which code version was actually running and buggered things up and commingled the versions instead, eh?  ..... I have changed it back in the latest release.

Zendog committed:
Added artwork processing

Added artwork metadata media server processing

Zendog committed:
Modified Grouped Artwork and State Processing

Changed where artwork is stored to historic indigo location. Aligned IP / Sonos Name / Indigo Device Names. Added menu function that dumps current group status for inspection and validation.

Zendog committed:
Tweaked metadata transition

Tweaked the transition from SiriusXM to Pandora to better align the artist, tack, album metadata if it is blank from XM transitioning.

Zendog committed:
Removed duplicated method

Removed duplicate artwork method and adjusted time pause for loop for re-driving artwork copy.

Zendog committed:
Update README.md

Zendog committed:
Update README.md

Zendog committed:
Update README.md

Zendog commented:
Newest version of the fork updates:

This version (1.0.2):

1. Updated SOCO library to - 0.30.9 for better control function support and overall stability.
2. Rewrote and fixed subscription services, includes fallback subscription upon failure to address devices falling of network.
3. Added menu option for dumping subscribed devices to the log as diagnostic aid on subscripition failures.
4. Rewrote discover to facilitate VLAN access for dedicated SONOS VLAN network access and control.
5. Rewrote and corrected volume controls.
6. Rewrote and corrected Bass and Treble controls.
7. Implemented SiriusXM login and channel processing.
8. Added menu options for dumping registered available XM Channel list to the log.
9. Rewrote and corrected Pandora loging / crypto processing from python 2 to python 3 based calls.
10. Implemented Pandora login and channel processing.
11. Added additional modules to requirements.txt.
12. Added Artwork / Metadata server.
13. Added menu option for dumping group device inventory / master controller identification to the log (matrixes Sonos SOCO Name / IP ADDRESS / Indigo Name / Indigo Device ID) for device management insight clarification.
14. Updated group controls to copy master controller enriched metadata states to all associated devices in the current device grouping.
15. This version additionally corrects all commented issues in the main branch.

Zendog committed:
Write default art as needed.

Write default artwork as needed in Sonos streaming if the saved JPG is too large or copy corrupted.

Zendog committed:
Merge branch 'main' of https://github.com/Zendog/Sonos

Zendog committed:
Removed PYC

Removed pycache and .pyc files that were left in in the last upload.

Zendog committed:
Added Source for control panel

Added Music Stream source variable for control panel use of Indigo current state for streaming sources being played.

Zendog committed:
Artwork compression test

Test pillow install to compress artwork processing

Zendog committed:
Artwork refinement

Artwork trigger enhancement and general event processing alignment.

Zendog committed:
Turned on slave copy

Turned on slave copy and reduced logging chatter.

Zendog committed:
XM Login Tracking / Group change blanks state and art

Enhanced xm login success / fail error checking. Blank out current state and art when group player goes standalone.

Zendog committed:
Cleaned libraries

Zendog committed:
Maintenance and Zoning 1

Maintenance release and 1st stage group zoning enabled

Zendog committed:
Turn off verbose logging

Zendog committed:
Test logical grouping synch

Testing for logical group device re-synch after group changes are processed.

Zendog committed:
Re-added plist file

Re-added plist file

Zendog committed:
Moved plist to contents directory

Moved plist to contents directory

Zendog committed:
Update Sonos.py

Updated grouping loops and general maintenance.

DaveL17 commented:
@Zendog are you ready for this fork to be merged into the Indigo-hosted repo? We had been holding off on that with your development work. Is it ready for a release to be generated?

Zendog commented:
Hello Dave,
May be best to wait a couple more days. Working on an update now to fix some lingering zone grouping issues. Will have a new release available soon.

Best regards,
Dean


> On Jul 31, 2025, at 06:06, Dave ***@***.***> wrote:
> 
> 
> DaveL17
>  left a comment 
> (IndigoDomotics/Sonos#12)
>  <https://github.com/IndigoDomotics/Sonos/pull/12#issuecomment-3139698969>
> @Zendog <https://github.com/Zendog> are you ready for this fork to be merged into the Indigo-hosted repo? We had been holding off on that with your development work. Is it ready for a release to be generated?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/IndigoDomotics/Sonos/pull/12#issuecomment-3139698969>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAT3FHO6IQCOHKL2LOWS6S33LIBEXAVCNFSM6AAAAAB34UG4KCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCMZZGY4TQOJWHE>.
> You are receiving this because you were mentioned.
>

DaveL17 commented:
Copy that. Let us know when you're ready.

Zendog committed:
Zone Groups Synchronization

Enhanced the zone group processing to detect drift and re-synch SOCO vs Plugin states.

Zendog commented:
I think we should be ready to commit the latest forked release now.

DT

> On Jul 31, 2025, at 18:01, Dave ***@***.***> wrote:
> 
> 
> DaveL17
>  left a comment 
> (IndigoDomotics/Sonos#12)
>  <https://github.com/IndigoDomotics/Sonos/pull/12#issuecomment-3141675152>
> Copy that. Let us know when you're ready.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/IndigoDomotics/Sonos/pull/12#issuecomment-3141675152>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAT3FHPMOX5QHMWD6ES7PCL3LKU5NAVCNFSM6AAAAAB34UG4KCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNBRGY3TKMJVGI>.
> You are receiving this because you were mentioned.
>

Zendog committed:
Enabled ActionDirect command s

Enabled volume and favorite actionDirect command processing

Zendog committed:
Beta Announcements

Beta testing announcements processing

Zendog committed:
Cleaned up action loops and ZGT

Fixed errors in action direct and zgt processing after announcement code fixes.

Zendog committed:
New version - Cleaned Up Startup

Cleaned startup grouping and debug logging verbosity.

Zendog committed:
Tweaked Group Cache

Tweaked the group states cache updates.

Zendog committed:
Changed to SOCO UUID Tracking

Changed grouped tracking to SOCO UUID Based.

Zendog committed:
Tweaked loop detection triggers

Increased loop detection trigger point to allow for additional rapid group changes.

Zendog committed:
IP Based Group Tracking

Trying to use IP based  cache index

Zendog committed:
Auto dump to log

Call dump_groups_to_log

Zendog committed:
Tweaked Dump To Log

Tweaked Dump To Log processing.

Zendog committed:
Tweaked Group Name

Initialize group name early for Londonmark script.

Zendog committed:
Monthly Maintenance

General bug fixes.

Zendog committed:
Announce state processing

Pre and Post state processing on annoucements.

Zendog committed:
Fix MP3 announcement staging and cache busting

Zendog committed:
Ignore cache and generated audio files

indigo-jay commented:
Hey all, I've lost the thread on this PR, it's been open for a while. @Zendog, I saw some commits to this PR today. A brief glance shows a `__pycache__` directory checked in which should not be there (should be in .gitignore probably) since those cache files are version specific. 

In any event, I have no hardware to test with nor do I know anything about the code. So after the removal of the `__pycache__` is it safe to merge and generate a release for the plugin store?

[EDIT] looks like there are merge conflicts, but just in the Info.plist which I can handle when doing the merge.

Zendog commented:
Yep … should be good to go. 

We may also want to align back to an actual forum area for the plugin support and comments. Tracking everything in one big “change of ownership” thread is hard to follow for the specific events.

Best regards,

Dean

> On Apr 30, 2026, at 15:03, Jay Martin ***@***.***> wrote:
> 
> 
> indigo-jay
>  left a comment 
> (IndigoDomotics/Sonos#12)
>  <https://github.com/IndigoDomotics/Sonos/pull/12#issuecomment-4356135792>
> Hey all, I've lost the thread on this PR, it's been open for a while. @Zendog <https://github.com/Zendog>, I saw some commits to this PR today. A brief glance shows a __pycache__ directory checked in which should not be there (should be in .gitignore probably) since those cache files are version specific.
> 
> In any event, I have no hardware to test with nor do I know anything about the code. So after the removal of the __pycache__ is it safe to merge and generate a release for the plugin store?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/IndigoDomotics/Sonos/pull/12#issuecomment-4356135792>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAT3FHICP5LNGQRBT3TMNZ34YO5RHAVCNFSM6AAAAACYMXQCLSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHM2DGNJWGEZTKNZZGI>.
> You are receiving this because you were mentioned.
>

indigo-jay committed:
Merge branch 'main' into main
